### PR TITLE
Add distributed MNIST training example for Kubernetes

### PR DESCRIPTION
**Pull request by fedebongio**
## **Summary**

Adds an end-to-end example under `examples/k8s/mnist-training/` demonstrating distributed MNIST training with JAX on TPU, orchestrated by a Kubernetes JobSet. This goes beyond the existing hello-world peer discovery scaffolding in `examples/k8s/` by showing actual data-parallel training with gradient averaging via `jax.lax.pmean`.

## **What's included**

* `train_mnist.py` — Pure JAX ConvNet with `shard_map` for data parallelism and `jax.distributed.initialize()` for automatic peer discovery on Kubernetes. Downloads MNIST directly from GCS with zero external data-loading dependencies.  
* `Dockerfile` — Minimal image with `jax[tpu]`, `jax[k8s]`, and optax.  
* `jobset.yaml` — JobSet manifest for TPU v5e training on GKE.  
* `svc-acct.yaml` — RBAC ServiceAccount for JAX's pod-based peer discovery.  
* `README.md` — Setup instructions, architecture explanation, GPU adaptation guide, and scaling tips (multislice, Kueue).

## **Design choices**

* **Pure JAX, no Flax:** keeps the example self-contained with zero framework dependencies beyond JAX \+ optax.  
* **`shard_map` \+ global `Mesh`:** uses the modern JAX sharding API rather than the older `pmap` pattern. `jax.device_put` with `NamedSharding` handles data distribution across processes automatically.  
* **JobSet only, no Kueue:** minimal orchestration footprint. The README documents Kueue as a scaling-up option.  
* **Zero-dependency data loading:** downloads MNIST directly from GCS via `urllib`, avoiding `tensorflow_datasets` and `keras` dependencies that add image size and can cause version conflicts.

## **Testing**

* **Single TPU VM** (standalone, no Kubernetes): trained to 98.6% test accuracy in 5 epochs. Validates model correctness, data loading, and TPU execution independently of the orchestration layer.  
* **GKE Autopilot cluster** with TPU v5e (2x4 topology, JobSet with 2 workers): validated distributed initialization, peer discovery via `jax[k8s]`, and multi-process gradient averaging.

```
│ [Process 0] Local devices: [TpuDevice(id=0, process_index=0, coords=(0,0,0), core_on_chip=0), TpuDevice(id=1, process_index=0, coords=(1,0,0),   │
│ core_on_chip=0), TpuDevice(id=2, process_index=0, coords=(0,1,0), core_on_chip=0), TpuDevice(id=3, process_index=0, coords=(1,1,0),              │
│ core_on_chip=0)]
  Final Training Results:
   - Epoch 5/5 loss: 0.0375
   - Test accuracy: 98.8%
```

### CODE REVIEW

File: examples/k8s/mnist-training/train_mnist.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        sharding = NamedSharding(mesh, PartitionSpec('batch'))
+        batch_images = jax.device_put(batch_images, sharding)
+        batch_labels = jax.device_put(batch_labels, sharding)
+
+        with mesh:
+            correct += int(eval_step(params, batch_images, batch_labels, mesh))
+
+    if process_idx == 0:
+        accuracy = correct / len(test_images) * 100
+        print(f"Test accuracy: {correct}/{len(test_images)} ({accuracy:.1f}%)")
```
[the remark this answers is not in the document]
Comment (fedebongio) [reply examples/k8s/mnist-training/train_mnist.py]: You are right, thanks. Fixed.
